finite_difference/src/subdomain_mod.f90

```diff
@@ -0,0 +1,302 @@
+    !! number of MPI ranks (or 1 if serial).
+    integer, intent(in), optional :: ndomains
+    !> Optional specification of the dimensions of the tiling grid
+    integer, intent(in), optional :: ndomainx, ndomainy
+    !> Width of halo to allow for when constructing sub-domains.
+    !! Default value is 1. Must be > 0 for a multi-processor run.
+    integer, intent(in), optional :: halo_width
+    ! Locals
+    integer :: ival, jval ! For tile extent calculation
+    integer :: internal_width, internal_height
+    integer :: ierr, nwidth
+    integer :: ji,jj, ith
+    integer :: jover, junder, height
+    integer :: iover, iunder, width
+    ! For doing stats on tile sizes
+    integer :: nvects, nvects_sum, nvects_min, nvects_max 
+    logical, parameter :: print_tiles = .TRUE.
+    ! Whether to automatically compute the dimensions of the tiling grid
+    logical :: auto_tile = .TRUE.
+    integer :: xlen, ylen
+    integer :: ntilex, ntiley
+    ! Local var to hold number of sub-domains
+    integer :: ndom = 1
+    ! Local var to hold halo width
+    integer :: hwidth = 1
+    ! Pointer to the current subdomain being defined
+    type(subdomain_type), pointer :: subdomain
+
+    ! Deal with optional arguments to this routine
+    if(.not. present(ndomains))then
+       if(.not. present(ndomainx) .and. .not. present(ndomainy))then
+          ! Automatically use the number of MPI processes
+          ndom = get_num_ranks()
+          auto_tile = .TRUE.
+       else if(present(ndomainx) .and. present(ndomainy))then
+          ndom = ndomainx * ndomainy
+          auto_tile = .FALSE.
+       else
+          call parallel_abort('decompose: invalid arguments supplied')
+       end if
+    else
+       ndom = ndomains
+       auto_tile = .TRUE.
+    end if
+
+    if(present(halo_width))then
+       if(halo_width < 1 .and. get_num_ranks() > 1)then
+          call parallel_abort('decompose: halo width must be > 0 if '// &
+                              'running on more than one process')
+       end if
+       hwidth = halo_width
+    end if
+
+    decomp%ndomains = ndom
+    allocate(decomp%subdomains(ndom), Stat=ierr)
+    if(ierr /= 0)then
+       call parallel_abort('decompose: failed to allocate tiles array')
+    end if
+
+    xlen = domainx
+    ylen = domainy
+    
+    if(auto_tile)then
+
+       ntilex = INT( SQRT(REAL(ndom)) )
+       do WHILE(MOD(ndom, ntilex) /= 0)
+          ntilex = ntilex - 1
+       end do
+       ntiley = ndom/ntilex
+
+       ! Match longest dimension of domain to longest dimension of 
+       ! process grid
+       if(xlen > ylen)then
+          if( ntilex < ntiley )then
+             ierr   = ntiley
+             ntiley = ntilex
+             ntilex = ierr
+          end if
+       else
+          ! N >= M so want nprocy >= nprocx
+          if( ntiley < ntilex )then
+             ierr   = ntiley
+             ntiley = ntilex
+             ntilex = ierr
+          end if
+       end if
```

Comment:
Impressive attention to detail :+1: I would have forced the user to supply the domain decomposition, but this is certainly useful - well done!

Reply:
Thanks :-) It was some code I had "prepared earlier" (do you have Blue Peter in Oz?)

Reply:
Oh well. Kid's TV show where they do a craft activity and then skip to the finished article with a "here's one I prepared earlier".